### PR TITLE
[feature] Self-serve email change for users

### PR DESCRIPTION
**Pull request by tsmethurst**
# Description

> If this is a code change, please include a summary of what you've coded, and link to the issue(s) it closes/implements.
>
> If this is a documentation change, please briefly describe what you've changed and why.

This pull request implements self-serve email address changes, so that admins don't have to manually change people's addresses in the database anymore.

To support this, adds two new endpoints: `GET /api/v1/user` and `POST /api/v1/user/email_change`.

UI preview of the user email change thingy in the settings panel

Closes https://github.com/superseriousbusiness/gotosocial/issues/2373

## Checklist

Please put an x inside each checkbox to indicate that you've read and followed it: `[ ]` -> `[x]`

If this is a documentation change, only the first checkbox must be filled (you can delete the others if you want).

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have not leveraged AI to create the proposed changes.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [ ] I/we have made any necessary changes to documentation.
- [ ] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [ ] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: internal/processing/user/email.go [hunk 1]
```diff
@@ -23,11 +23,92 @@ import (
...
 	"github.com/superseriousbusiness/gotosocial/internal/gtsmodel"
+	"github.com/superseriousbusiness/gotosocial/internal/messages"
+	"github.com/superseriousbusiness/gotosocial/internal/validate"
+	"golang.org/x/crypto/bcrypt"
 )
 
+// EmailChange processes an email address change request for the given user.
+func (p *Processor) EmailChange(
+	ctx context.Context,
+	user *gtsmodel.User,
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/user/email.go:39]: do we need to accept both user and account here given that the populated user contains its related account?

Comment (NyaaaWhatsUpDoc) [reply internal/processing/user/email.go:39]: or worst case can just do a populate step beforehand. just reduces (though obviously not totally prevents) risk of different account to the user being passed

File: internal/processing/user/email.go [hunk 2]
```diff
@@ -23,11 +23,92 @@ import (
...
+	if err := p.state.DB.UpdateUser(
+		ctx, user,
+		"unconfirmed_email",
+	); err != nil {
+		err := gtserror.Newf("db error updating user: %w", err)
+		return nil, gtserror.NewErrorInternalError(err)
+	}
+
+	// If account isn't set on User yet, do it here.
+	if user.Account == nil {
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/user/email.go:93]: given the above comment about function arguments this then won't be needed :)

File: internal/processing/workers/fromclientapi.go [hunk 3]
```diff
@@ -71,9 +71,9 @@ func (p *Processor) ProcessFromClientAPI(ctx context.Context, cMsg *messages.Fro
 	case ap.ActivityCreate:
 		switch cMsg.APObjectType {
 
-		// CREATE PROFILE/ACCOUNT
-		case ap.ObjectProfile, ap.ActorPerson:
-			return p.clientAPI.CreateAccount(ctx, cMsg)
+		// CREATE USER (ie., new user sign-up)
+		case ap.ActorPerson:
```
Comment (NyaaaWhatsUpDoc) [inline internal/processing/workers/fromclientapi.go:75]: shouldn't this be `ObjectProfile` given it's acting on a user? and same goes for the Accept / Reject account (though maybe rename those to user) functions below?

**Comment by NyaaaWhatsUpDoc** [comment]
a few comments but otherwise looks good :)

File: internal/processing/workers/fromclientapi.go (hunk 3 above)
Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go:75]: Ah yes good one!

Comment (tsmethurst) [reply internal/processing/workers/fromclientapi.go]: Wrote this to try to more clearly separate between account (remote account, local account where user is not involved/relevant) and user+account (always local, where user is involved as well): https://github.com/superseriousbusiness/gotosocial/pull/2957/commits/3742a1d6765ddfe5aed68d746092a70e5b369dc6

File: internal/processing/user/email.go (hunk 1 above)
Comment (tsmethurst) [reply internal/processing/user/email.go:39]: Done! https://github.com/superseriousbusiness/gotosocial/pull/2957/commits/2a6971b5f5c0e5c9073669fd3d4f8d7c3d8bda45

**Pull request merged**